src/hotspot/share/gc/shared/referenceProcessor.cpp

```diff
@@ -516,105 +512,131 @@ size_t ReferenceProcessor::total_reference_count(ReferenceType type) const {
+            closure_context),
+    _policy(policy) { }
+
+  virtual void work(uint worker_id)
   {
+    ResourceMark rm;
     RefProcSubPhasesWorkerTimeTracker tt(ReferenceProcessor::SoftRefSubPhase1, _phase_times, worker_id);
     size_t const removed = _ref_processor.process_soft_ref_reconsider_work(_ref_processor._discoveredSoftRefs[worker_id],
                                                                            _policy,
-                                                                           &is_alive,
-                                                                           &keep_alive,
-                                                                           &complete_gc);
+                                                                           _closure_context.is_alive(worker_id),
+                                                                           _closure_context.keep_alive(worker_id),
+                                                                           _closure_context.complete_gc(worker_id));
     _phase_times->add_ref_cleared(REF_SOFT, removed);
   }
 private:
   ReferencePolicy* _policy;
 };
 
-class RefProcPhase2Task: public AbstractRefProcTaskExecutor::ProcessTask {
+class RefProcPhase2Task: public RefProcTask {
   void run_phase2(uint worker_id,
                   DiscoveredList list[],
-                  BoolObjectClosure& is_alive,
-                  OopClosure& keep_alive,
+                  AbstractClosureContext& closure_context,
                   bool do_enqueue_and_clear,
                   ReferenceType ref_type) {
     size_t const removed = _ref_processor.process_soft_weak_final_refs_work(list[worker_id],
-                                                                            &is_alive,
-                                                                            &keep_alive,
+                                                                            closure_context.is_alive(worker_id),
+                                                                            closure_context.keep_alive(worker_id),
                                                                             do_enqueue_and_clear);
     _phase_times->add_ref_cleared(ref_type, removed);
   }
 
 public:
   RefProcPhase2Task(ReferenceProcessor& ref_processor,
-                    ReferenceProcessorPhaseTimes* phase_times)
-    : ProcessTask(ref_processor, false /* marks_oops_alive */, phase_times) { }
+                    ReferenceProcessorPhaseTimes* phase_times,
+                    AbstractClosureContext& closure_context)
+    : RefProcTask("RefProcPhase2Task",
+              ref_processor,
+              phase_times,
+              closure_context) { }
 
-  virtual void work(uint worker_id,
-                    BoolObjectClosure& is_alive,
-                    OopClosure& keep_alive,
-                    VoidClosure& complete_gc) {
+  virtual void work(uint worker_id) {
+    ResourceMark rm;
```

Comment:
The `ResourceMark`s was used in _some_ of the previous nested closures. See removal of  `G1STWRefProcTaskProxy` and `G1CMRefProcTaskProxy`.

Reply:
Yes, and I'm not sure why any of them exist, other than some paranoia that there might be some not properly scoped resource allocations somewhere inside that we want to squash on the way out of the thread.  That is, these seem like just hiding potential bugs.  But you probably don't want to chase down any actual bugs of that sort as part of this change, so okay, leave them.

Reply:
Hi, and sorry for the long silence from me.

I got feedback from Albert on improving the structure of this change and to use proxy classes instead of ClosureContexts. This will solve the problem with heap allocated closures and also improve readability IMO.

The Change is big, so I will not push it to this review thread before you agree with that it is the right way to go, but you can see the change here: https://github.com/lkorinth/jdk/commit/5cdddacd037c7ea4baf648262b9637540c5ec3e4

If you agree with me, I will push that change to this review.